mobile-ui/src/main/java/org/sagebionetworks/research/mobile_ui/recorder/distance/DistanceRecorder.java

```diff
@@ -0,0 +1,279 @@
+            if (this.getLastLocation() != null) {
+                long timeChangeMicros = nextLocation.getTime() - this.getLastLocation().getTime();
+                float distanceChangeMeters = this.getLastLocation().distanceTo(nextLocation);
+                this.currentSpeed = distanceChangeMeters / (timeChangeMicros * 1e-6);
+            }
+
+            super.update(nextLocation);
+        }
+
+        /**
+         * Returns the most recently measured speed of the user.
+         * @return the most recently measured speed of the user.
+         */
+        public double getCurrentSpeed() {
+            return this.currentSpeed;
+        }
+    }
+
+    /**
+     * Subscribes to the same Location stream as the DistanceRecorder and maintains a Summary of the recorder's
+     * recordings that will complete when the recorder is finished. This class should be used when the recorder's
+     * findings are only needed after the recorder is done recording.
+     */
+    protected static class DistanceSummarySubscriber implements SummarySubscriber<DistanceInfo, Location>,
+            SingleOnSubscribe<DistanceInfo> {
+        protected Set<SingleEmitter<DistanceInfo>> observers;
+        protected DistanceInfo result;
+
+        public DistanceSummarySubscriber() {
+            this.observers = new HashSet<>();
+            this.result = new DistanceInfo();
+        }
+
+        @Override
+        public Single<DistanceInfo> getSummary() {
+            return Single.create(this);
+        }
+
+        @Override
+        public void onSubscribe(final Subscription s) {
+            s.request(Long.MAX_VALUE);
+        }
+
+        @Override
+        public void onNext(final Location location) {
+            this.result.update(location);
+        }
+
+        @Override
+        public void onError(final Throwable t) {
+            for (SingleEmitter<DistanceInfo> emitter : this.observers) {
+                emitter.onError(t);
+            }
+        }
+
+        @Override
+        public void onComplete() {
+            for (SingleEmitter<DistanceInfo> emitter : this.observers) {
+                emitter.onSuccess(this.result);
+            }
+        }
+
+        @Override
+        public void subscribe(final SingleEmitter<DistanceInfo> emitter) throws Exception {
+            this.observers.add(emitter);
+        }
+    }
+
+    /**
+     * Subscribes to the same Location stream as the DistanceRecorder and maintains and provides access to the current
+     * state of the recorder. This class should be used when the recorders readings are needed in real time.
+     */
+    protected static class DistanceStateSubscriber implements CurrentStateSubscriber<CurrentDistanceInfo, Location> {
+        protected CurrentDistanceInfo result;
+
+        public DistanceStateSubscriber() {
+            this.result = new CurrentDistanceInfo();
+        }
+
+        @Override
+        public CurrentDistanceInfo getCurrentState() {
+            return this.result;
```

Comment:
Doesn't seem like this would be threadsafe
Any reason we shouldn't use something like `locationObservable.scan(locationAccumulator)`